pkg/storage/read_as_of_iterator.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package storage
+
+import "github.com/cockroachdb/cockroach/pkg/util/hlc"
+
+// ReadAsOfIterator wraps an iterator that implements the SimpleMVCCIterator
+// such that all MVCC keys returned are not delete tombstones and will have
+// timestamps less than the endtime timestamp, unless that key is invalid.
+type ReadAsOfIterator struct {
+	iter SimpleMVCCIterator
+
+	// endtime is the latest timestamp an MVCC key will have that the
+	// ReadAsOfIterator will return
+	endTime hlc.Timestamp
+
+	// valid is the latest boolean value returned by Valid()
+	valid bool
+}
+
+var _ SimpleMVCCIterator = &ReadAsOfIterator{}
+
+// Close no-ops. The caller is responsible for closing the underlying iterator.
+func (f *ReadAsOfIterator) Close() {
+}
+
+// SeekGE advances the iterator to the first key in the engine which is >= the
+// provided key and less than f.endtime.
+//
+// If Valid() is false after SeekGE() returns, the iterator may be at an invalid
+// key with a timestamp greater than or equal to f.endtime.
+func (f *ReadAsOfIterator) SeekGE(key MVCCKey) {
+	f.iter.SeekGE(key)
+	f.advance()
+}
+
+// Valid must be called after any call to Seek(), Next(), or similar methods. It
```

Comment:
nit: for most interface methods that don't do anything out of the ordinary, it's better to just say "Valid implements SimpleMVCCIterator." so we don't have to update a ton of comments if we ever change anything.

Reply:
Done.